src/listener_registry.rs

```diff
@@ -0,0 +1,314 @@
+use tokio::task::LocalSet;
+use tokio::time::Instant;
+use up_rust::{ComparableListener, UListener, UUri};
+use up_rust::{UCode, UMessage, UStatus};
+use vsomeip_proc_macro::generate_message_handler_extern_c_fns;
+use vsomeip_sys::glue::{make_application_wrapper, make_message_wrapper, make_runtime_wrapper};
+use vsomeip_sys::safe_glue::get_pinned_runtime;
+use vsomeip_sys::vsomeip;
+
+const THREAD_NUM: usize = 10;
+
+// Create a separate tokio Runtime for running the callback
+lazy_static! {
+    static ref CB_RUNTIME: Runtime = tokio::runtime::Builder::new_multi_thread()
+        .worker_threads(THREAD_NUM)
+        .enable_all()
+        .build()
+        .expect("Unable to create callback runtime");
+}
+
+static RUNTIME: Lazy<Arc<Runtime>> =
+    Lazy::new(|| Arc::new(Runtime::new().expect("Failed to create Tokio runtime")));
+
+fn get_runtime() -> Arc<Runtime> {
+    Arc::clone(&RUNTIME)
+}
+
+generate_message_handler_extern_c_fns!(10000);
+
+type ListenerIdMap = RwLock<HashMap<(UUri, Option<UUri>, ComparableListener), usize>>;
+lazy_static! {
+    static ref LISTENER_ID_CLIENT_ID_MAPPING: RwLock<HashMap<usize, ClientId>> =
+        RwLock::new(HashMap::new());
+    static ref LISTENER_ID_AUTHORITY_NAME: RwLock<HashMap<usize, AuthorityName>> =
+        RwLock::new(HashMap::new());
+    static ref LISTENER_ID_REMOTE_AUTHORITY_NAME: RwLock<HashMap<usize, AuthorityName>> =
+        RwLock::new(HashMap::new());
+    static ref LISTENER_REGISTRY: RwLock<HashMap<usize, Arc<dyn UListener>>> =
+        RwLock::new(HashMap::new());
+    static ref LISTENER_ID_MAP: ListenerIdMap = RwLock::new(HashMap::new());
+}
+
+pub(crate) async fn free_listener_id(listener_id: usize) -> UStatus {
+    info!("listener_id was not used since we already have registered for this");
+    let mut free_ids = FREE_LISTENER_IDS.write().await;
+    free_ids.insert(listener_id);
+    UStatus::fail_with_code(
+        UCode::ALREADY_EXISTS,
+        "Already have registered with this source, sink and listener",
+    )
+}
+
+pub(crate) async fn insert_into_listener_id_map(
+    authority_name: &AuthorityName,
+    remote_authority_name: &AuthorityName,
+    key: (UUri, Option<UUri>, ComparableListener),
+    listener_id: usize,
+) -> bool {
+    trace!(
+        "authority_name: {}, remote_authority_name: {}, listener_id: {}",
+        authority_name,
+        remote_authority_name,
+        listener_id
+    );
+
+    // TODO: Should ensure that we don't record a partial transaction by rolling back any pieces which succeeded if a latter part fails
+
+    let mut id_map = LISTENER_ID_MAP.write().await;
+    if id_map.insert(key, listener_id).is_some() {
+        trace!(
+            "Not inserted into LISTENER_ID_MAP since we already have registered for this Request"
+        );
+        return false;
+    } else {
+        trace!("Inserted into LISTENER_ID_MAP");
+    }
+
+    let mut listener_id_authority_name = LISTENER_ID_AUTHORITY_NAME.write().await;
```

Comment:
I think there shouldnt be any issues with the write locks as they are written right now but maybe consider clustering all of the write locks intended to be used in the function at the top of the function so that it is easier to see what locks are being used since the lock lasts for the duration of the function block?

Reply:
Yes, this section of code does need a bit of a cleanup, I feel. Thanks for pointing it out!

Reply:
I cleaned this up quite bit, feel free to look again.